pkg/controller/subjectpermission/subjectpermission_controller.go

```diff
@@ -171,18 +171,84 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
 		return reconcile.Result{}, nil
 	}
 
+	// get the NamespaceList
+	nsList := &corev1.NamespaceList{}
+	opts = client.ListOptions{Namespace: request.Namespace}
+	err = r.client.List(context.TODO(), &opts, nsList)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get clusterRoleBindingList")
+		return reconcile.Result{}, err
+	}
+
+	// slice of clusterRoleName that does not exists as a clusterRole
+	permissionClusterRoleNameList := controllerutil.PopulateCrPermissionClusterRoleNames(instance, clusterRoleList)
+	for _, permissionClusterRoleName := range permissionClusterRoleNameList {
+		// update condition
+		updatedSubjectPermission := controllerutil.UpdateCondition(instance, permissionClusterRoleName+" for clusterPermission does not exist", permissionClusterRoleName, true, managedv1alpha1.SubjectPermissionFailed)
+		err = r.client.Status().Update(context.TODO(), updatedSubjectPermission)
+		if err != nil {
+			reqLogger.Error(err, "Failed to update condition.")
+			return reconcile.Result{}, err
+		}
+	}
+
+	// compile list of allowed namespaces only for this subject permission. NOT a list of subject permissions
+	for _, permission := range instance.Spec.Permissions {
+		// list of all namespaces in safelist
+		safeList := controllerutil.GenerateSafeList(permission.NamespacesAllowedRegex, permission.NamespacesDeniedRegex, nsList)
+
+		// for each safelisted namespace
+		for _, ns := range safeList {
+			// get a list of all rolebindings in namespace
+			rbList := &v1.RoleBindingList{}
+			opts := client.ListOptions{Namespace: ns}
+			err = r.client.List(context.TODO(), &opts, rbList)
+
+			// create roleBinding
+			roleBinding := controllerutil.NewRoleBindingForClusterRole(permission.ClusterRoleName, instance.Spec.SubjectName, instance.Spec.SubjectKind, ns)
+			err := r.client.Create(context.TODO(), roleBinding)
+			if err != nil {
+				// if the rolebinding already exists then exit
+				for _, rb := range rbList.Items {
+					if roleBinding.Name == rb.Name {
+						break
+					}
+				}
+				// update the condition
+				permissionUpdatedCondition := controllerutil.UpdateCondition(instance, "Unable to create RoleBinding: "+err.Error(), permission.ClusterRoleName, true, managedv1alpha1.SubjectPermissionFailed)
+				err = r.client.Status().Update(context.TODO(), permissionUpdatedCondition)
+				if err != nil {
+					reqLogger.Error(err, "Failed to update condition.")
+					return reconcile.Result{}, err
+				}
+				reqLogger.Error(err, "Failed to create clusterRoleBinding")
+				return reconcile.Result{}, err
+			}
+
+			// update condition
+			permissionUpdatedCondition := controllerutil.UpdateCondition(instance, "Succesfully created RoleBinding", permission.ClusterRoleName, true, managedv1alpha1.SubjectPermissionCreated)
```

Comment:
I feel like all the update condition logic happens a lot too. Could this be made into a function? Including the err checking and reconcile exit.